core/src/main/java/com/radixdlt/consensus/liveness/Pacemaker.java

```diff
@@ -280,82 +219,165 @@ public void processBFTUpdate(BFTInsertUpdate update) {
+
+      // 1. Check if we haven't yet voted for anything (if so, ignore)
+      if (this.safetyRules.getLastVote(currentRound()).isPresent()) {
+        return;
+      }
+
+      /* 2. If we haven't yet voted, try to vote for whatever vertex was inserted
+      We only expect three kinds of vertices here:
+       a) a vertex received in a proposal for the current round: "normal" scenario
+       b) a vertex received in a QC for the next round: a scenario when there's already a QC
+          for the current round being processed but the Pacemaker hasn't received it yet.
+          In this case we might still want to send our (obsolete) vote
+       c) a fallback vertex has been inserted: in this case the Pacemaker's `roundStatus`
+          must have already been set to either PROPOSAL_REJECTED or TIMED_OUT
+
+      Since in all 3 cases we want to attempt a vote (if this is our first vote), we're just passing
+      over to `attemptVoteOnVertex` which takes care of deciding whether a vote should include
+      a timeout flag and whether it should be sent to the next leader or broadcasted to everyone. */
+      attemptVoteOnVertex(vertex);
+    } else if (round.gt(currentRound())) {
+      // A vertex for a future round (from Pacemaker's point of view) has been inserted
+      // this might be due to a race condition between this event and a RoundUpdate.
+      // We store the vertex so that the Pacemaker can use it once it switches to the next round
+      // (see `startRound`).
+      this.insertedVertexCarriedOverFromPrevRound = Optional.of(vertex);
+    } else {
       log.trace(
           "InsertUpdate: Ignoring insert {} for round {}, current round at {}",
           update,
           round,
-          this.latestRoundUpdate.getCurrentRound());
-      return;
+          currentRound());
     }
+  }
+
+  private void attemptVoteOnVertex(ExecutedVertex executedVertex) {
+    final var bftHeader =
+        new BFTHeader(
+            executedVertex.getRound(),
+            executedVertex.getVertexHash(),
+            executedVertex.getLedgerHeader());
 
-    this.maybeLatestInsertUpdate = Optional.of(update);
+    final var maybeBaseVote =
+        this.safetyRules.createVote(
+            executedVertex.getVertexWithHash(),
+            bftHeader,
+            executedVertex.getTimeOfExecution(),
+            this.latestRoundUpdate.getHighQC());
 
-    final var updateIsInsertionOfTimeoutVertex =
-        this.timeoutVertexIdForCurrentRound
-            .filter(update.getInserted().getVertexHash()::equals)
-            .isPresent();
+    maybeBaseVote.ifPresentOrElse(
+        baseVote -> {
+          // A timeout flag is included if:
+          // a) any scheduled timeout event was observed (even if in the end the round was
+          // prolonged)
+          // b) a proposal has been rejected (which implicitly means that this must be a vote on a
+          // fallback vertex)
+          final var includeTimeoutFlag =
```

Comment:
Out of interest why do we include a timeout flag on the vote (to let people form a TC) in the "PROPOSAL_REJECTED" case? Isn't that already handled by voting on a fallback vertex?

I think it may be safer (from a 3-chain perspective) to hold back the timeout flag for if a timeout has occurred?

Reply:
hmh, yeah I guess that makes sense (3-chain perspective), but I feel like there was some reason why we've included the flag, which I can't recall now :P I guess it won't be that much of an issue once we implement "slight delay for processing TC"

Reply:
The concern was roughly if the timestamp was accepted by half of quorum but not the other half, we could end up without being able to form a QC, so we'd need to timeout - so why not include the timeout flag immediately.

But I didn't understand the nuances about fallback vertices and timeout certificates at that point. Essentially it doesn't help, because the timeout for votes on the proposed vertex are only going to be received after the timeout delay anyway.

All that happens is that it gives rise  for the ability of a single validator to cause a TC instead of a QC, before the round's actually timed out.

And causes slightly confusing logic here accounting for it :)

Reply:
haha, okay, fair enough :+1: